cmd/kube-proxy/app/server.go

```diff
@@ -197,7 +197,8 @@ func (o *Options) AddFlags(fs *pflag.FlagSet) {
 		"This parameter is ignored if a config file is specified by --config.")
 
 	fs.StringSliceVar(&o.config.NodePortAddresses, "nodeport-addresses", o.config.NodePortAddresses,
-		"A list of CIDR ranges that contain valid node IPs. If set, connections to NodePort services will only be accepted on node IPs in one of the indicated ranges. If unset, NodePort connections will be accepted on all local IPs. This parameter is ignored if a config file is specified by --config.")
+		"A list of CIDR ranges that contain valid node IPs. If set, connections to NodePort services will only be accepted on node IPs in one of the indicated ranges. If this is unset and --nodeport-addresses-primary is true, then NodePort connections will be accepted only on the node's primary IP(s). If this is unset and --nodeport-addresses-primary is false, then NodePort connections will be accepted on all local IPs. This parameter is ignored if a config file is specified by --config.")
+	fs.BoolVar(&o.config.NodePortAddressesPrimary, "nodeport-addresses-primary", o.config.NodePortAddressesPrimary, "Indicates that NodePort service connections should only be accepted on the node's primary IP(s) as indicated by the Node object. This must be false if --node-port-addresses is non-empty. This parameter is ignored if a config file is specified by --config.")
```

Comment:
Can you spec what it means to be "primary" more clearly?  In particular, the logic we use to determine that is, IMO, sketchy as heck.  Node.status.addresses is pretty under-specced.

Should we only be binding to the first address of each family (preferringi internal to external) or should we bind to all internal?  Or all internal and all external?

Reply:
The primary node IPs are the ones returned from [`utilnode.GetNodeHostIPs`](https://github.com/kubernetes/kubernetes/blob/v1.29.2/pkg/util/node/node.go#L67-L72), and are not just used by kube-proxy; that also defines what the `hostIPs` of a pod are, and the `podIPs` of a host-network pod.

The logic used to pick them is maybe not well-documented, but it has not changed in years:
- The primary node IP is the first InternalIP in node.status.addresses, or else the first ExternalIP in node.status.addresses if there are no InternalIPs. (If there are no ExternalIPs either then that's an error and pods will not get any hostIPs and probably lots of other things will fail.)
- If node.status.addresses is dual-stack, then the secondary node IP (or "the other primary node IP"? we're not very consistent about the naming of this one) is the first InternalIP of the opposite IP family from the primary node IP, or else the first ExternalIP of that family if there are no InternalIPs of that family.

> Should we only be binding to the first address of each family (preferringi internal to external) or should we bind to all internal? Or all internal and all external?

So, IIRC, the vsphere cloud provider copies all IPs from the primary NIC (or all NICs?) over to node.status.addresses, and in some scenarios, LoadBalancer IPs will be assigned to NICs, so we'd end up getting exactly the behavior we _didn't_ want.

The desired behavior for `--nodeport-addresses primary` is really "only bind to the IP that the cloud load balancer is going to use". So the hope is that using the primary node IP as defined above will implement that...

(And if it doesn't, you can still specify a different `--nodeport-addresses` manually...)

Reply:
Can we include some of that in the explanation?  E.g. "NodePort connections will be accepted only on the node's primary IP(s) (the first defined address of each IP family)"